### PR TITLE
Internal: Collect Kit Revert data into the usage data [ED-7426]

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: app/modules/import-export/usage.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+class Usage {
+
+	/**
+	 * Register hooks.
+	 *
+	 * @return void
+	 */
+	public function register() {
+		add_filter( 'elementor/tracker/send_tracking_data_params', function ( array $params ) {
```
Comment (mark-elementor) [inline app/modules/import-export/usage.php:19]: I would just split it into another function, to avoid the anonymous function.
```suggestion
		add_filter( 'elementor/tracker/send_tracking_data_params', [ $this, 'add_tracking_data' ] );
```

File: app/modules/import-export/module.php [hunk 2]
```diff
@@ -86,6 +86,8 @@ public function __construct() {
 		if ( ElementorUtils::is_wp_cli() ) {
 			\WP_CLI::add_command( 'elementor kit', WP_CLI::class );
 		}
+
+		( new Usage() )->register();
```
Comment (mark-elementor) [inline app/modules/import-export/module.php:90]: Not critical at all, but maybe rename "register()" to something more self-explanatory?

**Review by mark-elementor — CHANGES_REQUESTED** [review]

File: app/modules/import-export/usage.php (hunk 1 above)
Comment (ManorHazaz) [reply app/modules/import-export/usage.php:19]: That will force me to publicize my function, I prefer not to.

File: app/modules/import-export/module.php (hunk 2 above)
Comment (ManorHazaz) [reply app/modules/import-export/module.php:90]: Any suggestions?

File: app/modules/import-export/usage.php (hunk 1 above)
Comment (mark-elementor) [reply app/modules/import-export/usage.php:19]: Ok, no problem.

File: app/modules/import-export/module.php (hunk 2 above)
Comment (mark-elementor) [reply app/modules/import-export/module.php:90]: send_tracking?
send_tracking_data?
push_usage_data?
set_usage_data?

Comment (ManorHazaz) [reply app/modules/import-export/module.php]: Done

**Review by mark-elementor — DISMISSED** [review]

File: app/modules/import-export/module.php (hunk 2 above)
Comment (StyleShit) [reply app/modules/import-export/module.php]: I think `register` is fine. It's the method that registers the service actions (same as we did in the Notes)
See Laravel's `register` and `boot` methods for reference:
https://laravel.com/docs/9.x/providers#the-register-method
https://laravel.com/docs/9.x/providers#the-boot-method

File: app/modules/import-export/usage.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+class Usage {
+
+	/**
+	 * Register hooks.
+	 *
+	 * @return void
+	 */
+	public function set_usage_data() {
+		add_filter( 'elementor/tracker/send_tracking_data_params', function ( array $params ) {
+			$params['usages']['import_export']['revert'] = $this->get_revert_usage_data();
```
Comment (StyleShit) [inline app/modules/import-export/usage.php:20]: maybe make sure that `$params['usages']['import_export']` is an array before pushing into it (if maybe for some reason it'll run before the "main" IE usage data or something)

File: app/modules/import-export/usage.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	private function get_revert_usage_data() {
+		$revert_sessions = ( new Revert() )->get_revert_sessions();
+
+		$data = [];
+
+		foreach ( $revert_sessions as $revert_session ) {
+			$data[] = [
+				'kit_name' => $revert_session['kit_name'],
+				'source' => $revert_session['source'],
+				'revert_timestamp' => $revert_session['revert_timestamp'],
```
Comment (StyleShit) [inline app/modules/import-export/usage.php:40]: maybe cast to int as you did in the `total_time`?

File: app/modules/import-export/usage.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+		} );
+	}
+
+	/**
+	 * Get the Revert usage data.
+	 *
+	 * @return array
+	 */
+	private function get_revert_usage_data() {
+		$revert_sessions = ( new Revert() )->get_revert_sessions();
```
Comment (StyleShit) [inline app/modules/import-export/usage.php:32]: I think it should be a dependency of this class and shouldn't be instantiated here

File: tests/phpunit/elementor/app/import-export/processes/test-revert.php [hunk 6]
```diff
@@ -73,6 +73,8 @@ public function test_run__revert_all_one_imported_only() {
 		$this->assertEquals( $base_posts, $after_revert__posts );
 		$this->assertNotEquals( $after_import__posts, $after_revert__posts );
 
+		$this->assert_valid_revert_sessions_option();
```
Comment (StyleShit) [inline tests/phpunit/elementor/app/import-export/processes/test-revert.php:76]: Maybe pass the sessions as a parameter, because this way it feels obscure

Comment (StyleShit) [inline tests/phpunit/elementor/app/import-export/processes/test-revert.php:76]: how is this test related to the usage?

File: tests/phpunit/elementor/app/import-export/test-usage.php [hunk 7]
```diff
@@ -0,0 +1,61 @@
+<?php
+namespace Elementor\Tests\Phpunit\Elementor\App\ImportExport;
+
+use Elementor\App\Modules\ImportExport\Usage;
+use ElementorEditorTesting\Elementor_Test_Base;
+
+class Test_Usage extends Elementor_Test_Base {
+
+	public function setUp()
+	{
```
Comment (StyleShit) [inline tests/phpunit/elementor/app/import-export/test-usage.php:10]:
```suggestion
	public function setUp() {
```

File: tests/phpunit/elementor/app/import-export/test-usage.php [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+		remove_all_filters('elementor/tracker/send_tracking_data_params');
+
+		( new Usage() )->set_usage_data();
+	}
+
+	public function test_get_usage_data__without_any_data()	{
+		// Act.
+		$usage = apply_filters('elementor/tracker/send_tracking_data_params', []);
+
+		// Assert
```
Comment (StyleShit) [inline tests/phpunit/elementor/app/import-export/test-usage.php:22]:
```suggestion
		// Assert.
```

File: tests/phpunit/elementor/app/import-export/test-usage.php [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+		parent::setUp();
+
+		remove_all_filters('elementor/tracker/send_tracking_data_params');
+
+		( new Usage() )->set_usage_data();
+	}
+
+	public function test_get_usage_data__without_any_data()	{
+		// Act.
+		$usage = apply_filters('elementor/tracker/send_tracking_data_params', []);
```
Comment (StyleShit) [inline tests/phpunit/elementor/app/import-export/test-usage.php:20]:
```suggestion
		$usage = apply_filters('elementor/tracker/send_tracking_data_params', []);
```

PHPCS this whole file

File: tests/phpunit/elementor/app/import-export/test-usage.php [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+		// Assert
+		$import_export_usage = $usage['usages']['import_export'];
+
+		$this->assertEquals( [], $import_export_usage['revert'] );
+	}
+
+	public function test_get_usage_data__adds_usage_without_delete_existing() {
+		// Arrange.
+		update_option( 'elementor_revert_sessions', [
```
Comment (StyleShit) [inline tests/phpunit/elementor/app/import-export/test-usage.php:30]: do you have a const for that?

File: tests/phpunit/elementor/app/import-export/test-usage.php [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+			[
+				'kit_name' => 'Kit Name',
+				'source' => 'Source',
+				'revert_timestamp' => time(),
+				'import_timestamp' => time() - 100,
```
Comment (StyleShit) [inline tests/phpunit/elementor/app/import-export/test-usage.php:35]: just for OCDing, can you swap those so it'll be chronological? (import, then revert)

File: tests/phpunit/elementor/app/import-export/test-usage.php [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+			],
+		] );
+
+		$expected_usage_data = [
+			[
+				'kit_name' => 'Kit Name',
+				'source' => 'Source',
+				'revert_timestamp' => time(),
```
Comment (StyleShit) [inline tests/phpunit/elementor/app/import-export/test-usage.php:43]: maybe extract the `time()` to a variable (`$now`?) so if the test will take some time (for some reason...) it won't fail for this 1ms delay

File: tests/phpunit/elementor/app/import-export/test-usage.php [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+				'total_time' => 100,
+			],
+		];
+
+		// Act.
+		$usage = apply_filters( 'elementor/tracker/send_tracking_data_params', [
+			'usages' => [],
+			'something-that-should-stay' => 123,
```
Comment (StyleShit) [inline tests/phpunit/elementor/app/import-export/test-usage.php:51]: maybe also add `something-that-should-stay` to the usages array, because that's the one you actually modify

File: tests/phpunit/elementor/app/import-export/test-usage.php [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+		update_option( 'elementor_revert_sessions', [
+			[
+				'kit_name' => 'Kit Name',
+				'source' => 'Source',
+				'revert_timestamp' => time(),
+				'import_timestamp' => time() - 100,
+			],
+		] );
+
+		$expected_usage_data = [
```
Comment (StyleShit) [inline tests/phpunit/elementor/app/import-export/test-usage.php:39]: move that to the assert section

File: app/modules/import-export/module.php [hunk 15]
```diff
@@ -86,6 +86,8 @@ public function __construct() {
 		if ( ElementorUtils::is_wp_cli() ) {
 			\WP_CLI::add_command( 'elementor kit', WP_CLI::class );
 		}
+
+		( new Usage() )->set_usage_data();
```
Comment (StyleShit) [inline app/modules/import-export/module.php:90]:
```suggestion
		( new Usage() )->register();
```

See my comment in @mark-elementor's review comment

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: app/modules/import-export/usage.php (hunk 3 above)
Comment (ManorHazaz) [reply app/modules/import-export/usage.php:20]: We have talked about it, it will create all the array cells if needed :)

File: tests/phpunit/elementor/app/import-export/processes/test-revert.php (hunk 6 above)
Comment (ManorHazaz) [reply tests/phpunit/elementor/app/import-export/processes/test-revert.php:76]: It's not.
I just added this test into the revert since in the usage i'm using this option and the revert was not covering it with tests.

File: app/modules/import-export/usage.php (hunk 5 above)
Comment (ManorHazaz) [reply app/modules/import-export/usage.php:32]: Let's talk about it.

File: tests/phpunit/elementor/app/import-export/processes/test-revert.php (hunk 6 above)
Comment (ManorHazaz) [reply tests/phpunit/elementor/app/import-export/processes/test-revert.php:76]: Done

File: tests/phpunit/elementor/app/import-export/test-usage.php (hunk 7 above)
Comment (ManorHazaz) [reply tests/phpunit/elementor/app/import-export/test-usage.php:10]: Done

File: tests/phpunit/elementor/app/import-export/test-usage.php (hunk 8 above)
Comment (ManorHazaz) [reply tests/phpunit/elementor/app/import-export/test-usage.php:22]: Done

File: tests/phpunit/elementor/app/import-export/test-usage.php (hunk 9 above)
Comment (ManorHazaz) [reply tests/phpunit/elementor/app/import-export/test-usage.php:20]: giphy

File: tests/phpunit/elementor/app/import-export/test-usage.php (hunk 10 above)
Comment (ManorHazaz) [reply tests/phpunit/elementor/app/import-export/test-usage.php:30]: Done

File: tests/phpunit/elementor/app/import-export/test-usage.php (hunk 11 above)
Comment (ManorHazaz) [reply tests/phpunit/elementor/app/import-export/test-usage.php:35]: FUAFUOCD

File: tests/phpunit/elementor/app/import-export/test-usage.php (hunk 12 above)
Comment (ManorHazaz) [reply tests/phpunit/elementor/app/import-export/test-usage.php:43]: Nice one, Done

File: tests/phpunit/elementor/app/import-export/test-usage.php (hunk 13 above)
Comment (ManorHazaz) [reply tests/phpunit/elementor/app/import-export/test-usage.php:51]: Done

File: tests/phpunit/elementor/app/import-export/test-usage.php (hunk 14 above)
Comment (ManorHazaz) [reply tests/phpunit/elementor/app/import-export/test-usage.php:39]: Done

File: app/modules/import-export/module.php (hunk 15 above)
Comment (ManorHazaz) [reply app/modules/import-export/module.php:90]: Changed.
Mark is OK with both

**Review by StyleShit — APPROVED** [review]
